Ensure unique listener injection in TestNG

krmahadevan opened the pull request:
Fixes # .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

Fixes #1232

Altered TestNG logic to ensure that TestNG now 
adds only one instance per listener class so that
duplicate listener invocation can be prevented.

krmahadevan commented:
@juherr / @cbeust - Can you please help take a look at this and kindly help getting this merged ?

missedone commented:
I think it'd be better to print some warning message that multi listener of the same type found, this might help people diagnose this behavior change?

juherr commented on src/test/java/test/testng1232/TestListenerInstances.java:
```diff
@@ -0,0 +1,57 @@
+import test.SimpleBaseTest;
+import test.testng1232.TestListenerFor1232.CounterTypes;
+
+public class TestListenerInstances extends SimpleBaseTest {
```
It should be added in `testng.xml`

krmahadevan replied:
Oh yeah. I completely forgot about that. Thanks for reminding me.

juherr commented on src/main/java/org/testng/TestNG.java:
```diff
@@ -717,17 +717,28 @@ public void addListener(ITestNGListener listener) {
+      ISuiteListener suite = (ISuiteListener) listener;
+      if (! m_suiteListeners.containsKey(suite.getClass())) {
+        m_suiteListeners.put(suite.getClass(), (ISuiteListener) listener);
+      }
```
From @missedone

> I think it'd be better to print some warning message that multi listener of the same type found, this might help people diagnose this behavior change?

Agree!

krmahadevan replied:
Sure will do that.

krmahadevan replied:
Will take care.

juherr commented on src/main/java/org/testng/TestNG.java:
```diff
@@ -1368,11 +1376,11 @@ private SuiteRunner createSuiteRunner(XmlSuite xmlSuite) {
-        m_invokedMethodListeners,
-        m_testListeners,
-        m_classListeners);
+        Lists.newArrayList(m_invokedMethodListeners.values()),
```
We can change or add a new signature for `SuiteRunner` constructor with collections instead of lists. It will avoid useless objects creation.

juherr replied:
Sure. You can do it in 2 steps/commits: 1. fix it as minimal as possible, 2. Clean technical dept 👍

juherr commented on src/main/java/org/testng/TestNG.java:
```diff
@@ -2057,7 +2065,7 @@ public void setGroupByInstances(boolean b) {
 
   private URLClassLoader m_serviceLoaderClassLoader;
-  private List<ITestNGListener> m_serviceLoaderListeners = Lists.newArrayList();
+  private Map<Class<? extends ITestNGListener>, ITestNGListener> m_serviceLoaderListeners = Maps.newHashMap();
```
Could be `final`

juherr replied:
Everywhere, when possible. I think it is easier to read/understand when the file is big. It assures you the variable won't be changed somewhere.

juherr commented on src/test/java/test/testng1232/TestListenerInstances.java:
```diff
@@ -0,0 +1,57 @@
+        runTestForTestClass(TestClassContainer.SimpleTestClass.class, true);
+    }
+
+    private void runTestForTestClass(Class clazz) {
```
Could be static like the 2 methods above

juherr replied:
`Class<?>`

krmahadevan commented:
@juherr - I have fixed all the review comments. I have ran the tests locally and there are no failures. Not sure what is causing failures in Travis. The failures are sporadic and I am not able to figure out what test is failing in Travis CI so that I can dig in to see what is going on.

cbeust commented on src/main/java/org/testng/TestNG.java:
```diff
@@ -712,22 +712,33 @@ public void addListener(Object listener) {
 
+  private static <E> void wireListenerIfNotDuplicate(Map<Class<? extends E>, E> map, Class<? extends E> type, E value) {
+    if (map.containsKey(value.getClass())) {
+      LOGGER.warn("Skipping wiring in of the duplicate listener instance for the class : " + value.getClass().getName());
```
That's a bit arcane, how about simply "Ignoring duplicate listener %s"...?

krmahadevan replied:
Fixed this.

cbeust commented on src/main/java/org/testng/TestNG.java:
```diff
@@ -712,22 +712,33 @@ public void addListener(Object listener) {
     addListener((ITestNGListener) listener);
   }
 
+  private static <E> void wireListenerIfNotDuplicate(Map<Class<? extends E>, E> map, Class<? extends E> type, E value) {
```
Personally, I'd call it `maybeAddListener()`, but just nitpicking.

cbeust commented on src/main/java/org/testng/TestNG.java:
```diff
@@ -2070,14 +2072,16 @@ public void setServiceLoaderClassLoader(URLClassLoader ucl) {
    */
   private void addServiceLoaderListener(ITestNGListener l) {
-    m_serviceLoaderListeners.add(l);
+    if (! m_serviceLoaderListeners.containsKey(l.getClass())) {
```
Thinking out loud: now that we have all these listener maps scattered a bit everywhere, we might want to encapsulate this in a `ListenerContainer`. Have just one field of this type on this class and `m_listenerContainer.addServiceLoaderListener(...)`.

Don't do this in this PR, though (and not sure it's worth doing, just wondering).

krmahadevan commented:
@cbeust - I have addressed the review comments you shared. Please let me know if there is anything else from my side that's required to get this merged.

cbeust commented:
LGTM. @juherr?

juherr commented:
LGTM once travis is ok

krmahadevan commented:
@juherr - Out of the 3 flavors for which Travis builds, the JDK7 flavors are ending within a minute itself and there is no visibility into why its happening. The JDK8 build however runs to completion with all the tests passing.
At this point I am not sure as to what is causing the JDK7 build crashes. I have built the code locally and all tests run fine. I am not sure what else can be done from my side to get this issue with Travis sorted out!

juherr commented:
Did you find which tests are failing ?

krmahadevan commented:
Hard to say... All I see is 
```
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':test'.
> Process 'Gradle Test Executor 2' finished with non-zero exit value 137
* Try:
Run with --debug option to get more log output.

```
I don't think its even running all the tests in the suite xml file.

juherr commented:
Did you try to reproduce with jdk7 on a linux system? I'm not feeling good to merge something failing everytime on travis. Even more when we don't know what is happening.

krmahadevan commented:
@juherr - No I haven't. I have so far only built the code on a MAC. I will have to try and do that and then see what happens. I don't have a Linux machine at my disposal right now.

But just curious. How would that affect ? The builds are not even running fully for me to figure out what all tests ran and what didn't.

juherr commented:
Similar issues could help to find the fix: https://github.com/grpc/grpc-java/pull/1457 and https://github.com/mockito/mockito/pull/369

krmahadevan commented:
Yay!! @juherr  That suggestion did the trick. I finally have a green build with all the jdk flavor builds passed in Travis. Finally I think we can get this pull request merged.

@cbeust - Can you please help get this merged now ?

juherr commented:
:+1:

cbeust commented:
I went through the same Travis problem with Kobalt and JDK 7, the fix was also the memory flag.

LGTM, merging.
